xarray/tests/test_cftime_offsets.py

```diff
@@ -0,0 +1,723 @@
+    ('0001-01-01', None, 4, 'D', None, False,
+     [(1, 1, 1), (1, 1, 2), (1, 1, 3), (1, 1, 4)]),
+    (None, '0001-01-04', 4, 'D', None, False,
+     [(1, 1, 1), (1, 1, 2), (1, 1, 3), (1, 1, 4)]),
+    ((1, 1, 1), '0001-01-04', None, 'D', None, False,
+     [(1, 1, 1), (1, 1, 2), (1, 1, 3), (1, 1, 4)]),
+    ((1, 1, 1), (1, 1, 4), None, 'D', None, False,
+     [(1, 1, 1), (1, 1, 2), (1, 1, 3), (1, 1, 4)]),
+    ('0001-01-30', '0011-02-01', None, '3AS-JUN', None, False,
+     [(1, 6, 1), (4, 6, 1), (7, 6, 1), (10, 6, 1)]),
+    ('0001-01-04', '0001-01-01', None, 'D', None, False,
+     []),
+    ('0010', None, 4, YearBegin(n=-2), None, False,
+     [(10, 1, 1), (8, 1, 1), (6, 1, 1), (4, 1, 1)]),
+    ('0001-01-01', '0001-01-04', 4, None, None, False,
+     [(1, 1, 1), (1, 1, 2), (1, 1, 3), (1, 1, 4)])
+]
+
+
+@pytest.mark.skipif(not has_cftime, reason='cftime not installed')
+@pytest.mark.parametrize(
+    ('start', 'end', 'periods', 'freq', 'closed', 'normalize',
+     'expected_date_args'),
+    _CFTIME_RANGE_TESTS, ids=_id_func
+)
+def test_private_cftime_range(
+        start, end, periods, freq, closed, normalize, calendar,
+        expected_date_args):
+    date_type = get_date_type(calendar)
+    if isinstance(start, tuple):
+        start = date_type(*start)
+    if isinstance(end, tuple):
+        end = date_type(*end)
+    result = _cftime_range(
+        start, end, periods, freq, closed, normalize, calendar)
+    expected = [date_type(*args) for args in expected_date_args]
+    if freq is not None:
+        np.testing.assert_equal(result, expected)
+    else:
+        # If we create a linear range of dates using cftime.num2date
+        # we will not get exact round number dates.  This is because
+        # datetime arithmetic in cftime is accurate approximately to
+        # 1 millisecond (see https://unidata.github.io/cftime/api.html).
+        deltas = result - expected
+        deltas = np.array([delta.total_seconds() for delta in deltas])
+        assert np.max(np.abs(deltas)) < 0.001
+
+
+@pytest.mark.skipif(not has_cftime, reason='cftime not installed')
+@pytest.mark.parametrize(
+    ('start', 'end', 'periods', 'freq', 'closed'),
+    [(None, None, 5, 'A', None),
+     ('2000', None, None, 'A', None),
+     (None, '2000', None, 'A', None),
+     ('2000', '2001', None, None, None),
+     (None, None, None, None, None),
+     ('2000', '2001', None, 'A', 'up'),
+     ('2000', '2001', 5, 'A', None)]
+)
+def test_invalid_cftime_range_inputs(start, end, periods, freq, closed):
+    with pytest.raises(ValueError):
+        _cftime_range(start, end, periods, freq, closed=closed)
+
+
+@pytest.mark.skipif(not has_cftime, reason='cftime not installed')
+@pytest.mark.parametrize(
+    ('start', 'end', 'periods', 'freq', 'name'),
+    [('0001', None, 5, 'A', 'foo'),
+     ('2000', None, 5, 'A', 'foo'),
+     ('2000', '1999', None, 'A', 'foo')]
+)
+def test_cftime_range(start, end, periods, freq, name, calendar):
```

Comment:
Can you add selection of calendar specific tests, e.g., verify that generating generating a year's worth of dates with noleap, all_leap and 360_day does the right thing?